File: datastruct/tree/B-tree/src/btree.py
#!/usr/bin/python

# btree.py, B-Tree
# Copyright (C) 2010, Liu Xinyu (liuxinyu95@gmail.com)
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.

from random import sample, randint

TREE_2_3_4 = 2

class BTree:
    def __init__(self):
        self.keys = []
        self.subtrees = []

    def __str__(self):
        res = "("
        if is_leaf(self):
            res += ", ".join(str(k) for k in self.keys)
        else:
            for i in range(len(self.keys)):
                res += str(self.subtrees[i]) + ", " + str(self.keys[i]) + ", "
            res += str(self.subtrees[-1])
        return res + ")"

# d - 1 <= |t.keys| <= 2 * d - 1
#
# To avoid violating B-tree after insert/delete, when keys reach to either boundary
# (= hold), trigger split/merge

def is_leaf(t):
    return t.subtrees == []

def full(d, t):
    return len(t.keys) >= 2 * d - 1

def low(d, t):
    return len(t.keys) <= d - 1

def split(d, z, i):
    """split z.subtrees[i] at degree d"""
    x = z.subtrees[i]
    y = BTree()
    z.keys.insert(i, x.keys[d - 1])
    z.subtrees.insert(i + 1, y)
    y.keys = x.keys[d : ]
    x.keys = x.keys[ : d - 1]
    if not is_leaf(x):
        y.subtrees = x.subtrees[d : ]
        x.subtrees = x.subtrees[ : d]

def insert(d, t, x):
    """insert key x to tree t, where the degree is d"""
    root = t
    if full(d, root):
        s = BTree()
        s.subtrees = [root]
        split(d, s, 0)
        root = s
    return insert_nonfull(d, root, x)

def insert_nonfull(d, t, x):
    if is_leaf(t):
        ordered_insert(t.keys, x)
    else:
        i = len(t.keys)
        while i > 0 and x < t.keys[i-1]:
            i = i - 1
        if full(d, t.subtrees[i]):
            split(d, t, i)
            if x > t.keys[i]:
                i = i + 1
        insert_nonfull(d, t.subtrees[i], x)
    return t

def ordered_insert(lst, x):
    i = len(lst)
    lst.append(x)
    while i > 0 and lst[i] < lst[i - 1]:
        (lst[i - 1], lst[i]) = (lst[i], lst[i - 1])
        i = i - 1

def delete(d, t, x):
    if t.keys == []:
        return t
    for i in range(len(t.keys)):
        if x <= t.keys[i]:
            break
    if x == t.keys[i]:
        if is_leaf(t):         # case 1
            t.keys.pop(i)
        else:
            tl = t.subtrees[i]
            tr = t.subtrees[i + 1]
            if not low(d, tl):    # case 2a
                t.keys[i] = max_of(tl)
                delete(d, tl, t.keys[i])
            elif not low(d, tr):  # case 2b
                t.keys[i] = min_of(tr)
                delete(d, tr, t.keys[i])
            else:              # case 2c
                merge_subtrees(t, i)
                delete(d, tl, x)
                if t.keys == []:  # shrink height
                    t = tl
        return t
    if not is_leaf(t):
        if x > t.keys[-1]:
            i = i + 1
        if low(d, t.subtrees[i]):
            n = len(t.subtrees)
            tl = None if i == 0 else t.subtrees[i - 1]
            tr = None if i == n - 1 else t.subtrees[i + 1]
            if tl and (not low(d, tl)):   # case 3a, borrow from left
                t.subtrees[i].keys.insert(0, t.keys[i - 1])
                t.keys[i - 1] = tl.keys.pop()
                if not is_leaf(tl):
                    t.subtrees[i].subtrees.insert(0, tl.subtrees.pop())
            elif tr and (not low(d, tr)):  # case 3a, borrow from right
                t.subtrees[i].keys.append(t.keys[i])
                t.keys[i] = tr.keys.pop(0)
                if not is_leaf(tr):
                    t.subtrees[i].subtrees.append(tr.subtrees.pop(0))
            else:       # case 3b
                merge_subtrees(t, i if i + 1 < n else i - 1)
                if i + 1 == n:
                    i = i - 1
        delete(d, t.subtrees[i], x)
        if t.keys == []:    # shrink height
            t = t.subtrees[0]
    return t

def merge_subtrees(t, i):
    """merge t.subtrees[i], keys[i], t.subtrees[i+1]"""
    t.subtrees[i].keys += [t.keys[i]] + t.subtrees[i + 1].keys
    t.subtrees[i].subtrees += t.subtrees[i + 1].subtrees
    t.keys.pop(i)
    t.subtrees.pop(i + 1)

def max_of(t):
    while t.subtrees != []:
        t = t.subtrees[-1]
    return t.keys[-1]

def min_of(t):
    while t.subtrees != []:
        t = t.subtrees[0]
    return t.keys[0]

def lookup(t, k):
    if t.keys == []:
        return None
    for i in range(len(t.keys)):
        if k <= t.keys[i]:
            break
    if k == t.keys[i]:
        return (t, i)
    elif is_leaf(t):
        return None
    else:
        if k > t.keys[-1]:
            i = i + 1
        return lookup(t.subtrees[i], k)

def fromlist(d, xs):
    t = BTree()
    for x in xs:
        t = insert(d, t, x)
    return t

def tolist(t):
    xs = []
    if t is None:
        return xs
    if is_leaf(t):
        xs = [k for k in t.keys]
    else:
        for i in range(len(t.keys)):
            xs += tolist(t.subtrees[i])
            xs.append(t.keys[i])
        xs += tolist(t.subtrees[-1])
    return xs

# verification

def deg(xs):
    return (xs[0] % 5) + 2 if xs else 2

def is_btree(d, t, depth):
    if t is None:
        return True
    n = len(t.keys)
    if n > 2 * d - 1:
        return False
    if depth > 0 and n < d - 1:
        return False
    for tr in t.subtrees:
        if not is_btree(d, tr, depth + 1):
            return False
    return True

def prop_order(xs):
    t = fromlist(deg(xs), xs)
    ys = tolist(t)
    zs = sorted(xs)
    assert ys == zs, f"ys = {ys}, zs = {zs}, t = {t}"

def prop_insert(xs):
    d = deg(xs)
    t = fromlist(d, xs)
    assert is_btree(d, t, 0), f"violate B-tree: d = {d}, t = {t}"

def prop_lookup(xs):
    d = deg(xs)
    t = fromlist(d, xs)
    ys = sample(xs, min(5, len(xs))) + sample(range(100), 5)
    for y in ys:
        r = lookup(t, y)
        if y in xs:
            assert r, f"not found {y} in t = {t}"
            (tr, i) = r
            assert tr.keys[i] == y, f"y = {y}, tr = {tr}, i = {i}"
        else:
            assert (r is None), f"y = {y}, r = {r}"

def prop_delete(xs):
    d = deg(xs)
    ys = sample(xs, min(5, len(xs))) + sample(range(100), 5)
    for y in ys:
        t = fromlist(d, xs)
        r = delete(d, t, y)
        zs = tolist(r)
        assert is_btree(d, r, 0), f"violate B-tree: d = {d}, t={r}"
        if y in xs:
            assert y not in zs, f"found {y} after delete:\nt = {r}\nbefore delete\nt = {t}"
        else:
            assert sorted(xs) == zs, f"y = {y}, t = {r}, \nxs = {sorted(xs)}\nzs = {zs}"

def test(f):
    for _ in range(100):
        xs = sample(range(100), randint(0, 100))
        f(xs)
    print(f"100 tests for {f} passed.")

if __name__ == "__main__":
    test(prop_order)
    test(prop_insert)
    test(prop_lookup)
    test(prop_delete)

File: datastruct/tree/B-tree/src/BTree.hs
-- BTree.hs
-- Copyright (C) 2010, Liu Xinyu (liuxinyu95@gmail.com)

-- This program is free software: you can redistribute it and/or modify
-- it under the terms of the GNU General Public License as published by
-- the Free Software Foundation, either version 3 of the License, or
-- (at your option) any later version.

-- This program is distributed in the hope that it will be useful,
-- but WITHOUT ANY WARRANTY; without even the implied warranty of
-- MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
-- GNU General Public License for more details.

-- You should have received a copy of the GNU General Public License
-- along with this program.  If not, see <http://www.gnu.org/licenses/>.

module BTree where

-- Implement the B-Tree with List,
--   2-passes insert/delete (insert/delete then fix)

import qualified Data.List as L
import Test.QuickCheck
import Data.Maybe (listToMaybe)

data BTree a = BTree { keys :: [a]
                     , subTrees :: [BTree a] } deriving (Eq, Show)

empty = BTree [] []

-- Define BTree of degree d as (d, t) :: (Int, BTree a) where
--   d - 1 <= |keys t| <= 2 * d - 1

full d (BTree ks _) = (length ks) > 2 * d - 1

low d (BTree ks _) = (length ks) < d - 1

partition x (BTree ks ts) = (l, t, r) where
  l = (ks1, ts1)
  r = (ks2, ts2)
  (ks1, ks2) = L.span (< x) ks
  (ts1, (t:ts2)) = L.splitAt (length ks1) ts

split d (BTree ks ts) = (BTree ks1 ts1, k, BTree ks2 ts2) where
  (ks1, k:ks2) = L.splitAt (d - 1) ks
  (ts1, ts2) = L.splitAt d ts

unsplit (BTree ks1 ts1) k (BTree ks2 ts2) = BTree (ks1 ++ [k] ++ ks2)
                                                  (ts1 ++ ts2)

max' (BTree ks []) = last ks
max' (BTree _  ts) = max' $ last ts

insert :: (Ord a) => a -> (Int, BTree a) -> (Int, BTree a)
insert x (d, t) = fixRoot (d, ins t) where
    ins t@(BTree ks ts)
      | x `elem` ks = t -- skip duplicated key
      | null ts = BTree (L.insert x ks) []
      | otherwise = balance d l (ins t') r where
          (l, t', r) = partition x t

delete :: (Ord a, Eq a) => a -> (Int, BTree a) -> (Int, BTree a)
delete x (d, t) = fixRoot (d, del x t) where
    del x (BTree ks []) = BTree (L.delete x ks) []
    del x t = if (Just x) == listToMaybe ks' then
                let k' = max' t' in balance d l (del k' t') (k':(tail ks'), ts')
              else balance d l (del x t') r
      where
        (l, t', r@(ks', ts')) = partition x t

fixRoot (d, BTree [] [t]) = (d, t)
fixRoot (d, t) | full d t  = let (t1, k, t2) = split d t in
                               (d, BTree [k] [t1, t2])
               | otherwise = (d, t)

balance :: Int -> ([a], [BTree a]) -> BTree a -> ([a], [BTree a]) -> BTree a
balance d (ks1, ts1) t (ks2, ts2)
    | full d t  = fixFull
    | low  d t  = fixLow
    | otherwise = BTree (ks1 ++ ks2) (ts1 ++ [t] ++ ts2)
  where
    fixFull = let (t1, k, t2) = split d t in
                BTree (ks1 ++ [k] ++ ks2) (ts1 ++ [t1, t2] ++ ts2)
    fixLow | not $ null ks1 = balance d (init ks1, init ts1)
                                      (unsplit (last ts1) (last ks1) t)
                                      (ks2, ts2)
           | not $ null ks2 = balance d (ks1, ts1)
                                      (unsplit t (head ks2) (head ts2))
                                      (tail ks2, tail ts2)
           | otherwise = t -- ks2 == ks1 == []

lookup k t@(BTree ks []) = if k `elem` ks then Just t else Nothing
lookup k t = if (Just k) == listToMaybe ks then Just t
             else BTree.lookup k t'  where
  (_, t', (ks, _)) = partition k t

fromList d xs = foldr insert (d, empty) xs

toList (BTree ks []) = ks
toList (BTree ks (t:ts)) = concat ((toList t) : (zipWith (\k tr -> k : toList tr) ks ts))

foldt :: (a -> b -> b) -> b -> (BTree a) -> b
foldt f z (BTree ks []) = foldr f z ks
foldt f z (BTree [] [t]) = foldt f z t
foldt f z (BTree (k:ks) (t:ts)) = let z' = foldt f z (BTree ks ts) in
  foldt f (f k z') t

-- Quick check

degOf [] = 2
degOf (x:_) = x `mod` 5 + 2 -- limit degree: 2 <= d < 6 for simplicity

isBTree :: Int -> Int -> (BTree a) -> Bool
isBTree d n t@(BTree _ ts) = (not $ full d t)
                          && (n == 0 || (not $ low d t))
                          && (and [isBTree d (n + 1) tr | tr <- ts])

prop_order :: [Int] -> Bool
prop_order xs = (L.sort $ L.nub xs) == (toList $ snd $ fromList d xs) where
  d = degOf xs

prop_insert :: [Int] -> Bool
prop_insert xs = isBTree d 0 $ snd $ fromList d xs where d = degOf xs

prop_delete :: [Int] -> Int -> Bool
prop_delete xs x = (L.sort $ L.delete x ys) ==
                   (toList $ snd $ delete x $ fromList d ys) where
  ys = L.nub xs
  d = degOf xs

prop_del_balance :: [Int] -> Int -> Bool
prop_del_balance xs x = isBTree d 0 $ snd $ delete x $ fromList d ys where
  ys = L.nub xs
  d = degOf xs

prop_lookup :: [Int] -> Int -> Bool
prop_lookup xs x = f $ BTree.lookup x $ snd $ fromList d xs where
  d = degOf xs
  f Nothing = not $ elem x xs
  f (Just (BTree ks _)) = x `elem` ks

prop_foldt :: [Int] -> Bool
prop_foldt xs = toList t == foldt (:) [] t where
  t = snd $ fromList d xs
  d = degOf xs

testAll = do
  quickCheck prop_order
  quickCheck prop_insert
  quickCheck prop_delete
  quickCheck prop_del_balance
  quickCheck prop_lookup
  quickCheck prop_foldt
